flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/utils/ValuesOperationFactory.java

```diff
@@ -197,10 +199,18 @@ private Optional<ResolvedExpression> convertToExpectedType(
                     && targetLogicalType.is(ARRAY)) {
                 return convertArrayToExpectedType(
                         sourceExpression, (CollectionDataType) targetDataType, postResolverFactory);
-            } else if (functionDefinition == BuiltInFunctionDefinitions.MAP
-                    && targetLogicalType.is(MAP)) {
-                return convertMapToExpectedType(
-                        sourceExpression, (KeyValueDataType) targetDataType, postResolverFactory);
+            } else if (functionDefinition == BuiltInFunctionDefinitions.MAP) {
+                if (targetLogicalType.is(MAP)) {
```

Comment:
This piece of code confuses me. Why is the function `MAP` but the target type MULTISET? On an API level those two data types should be treated completely different.

Reply:
It seems the issue is that on api level `org.apache.flink.table.expressions.ApiExpressionUtils#objectToExpression` [1] translates maps to `MAP`. And both `MAP` and `MULTISET` are represented as maps there...  
[1] https://github.com/apache/flink/blob/e91eb5ec2fea9a2f28f3f55a06bc140e4ce4b5f5/flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/ApiExpressionUtils.java#L85-L104